pp_ctl.c - teach module_true how to deal with blocked requires

demerphq opened the pull request:
If require encounters an @INC hook that blocks the require it ends up with a STUB node as the optree for the LEAVEEVAL. This was causing the module_true logic to segfault. Guarding against the OP_STUB node fixes the problem.

This affected the tests for IO::Socket::SSL. Thanks to Graham Knopp for the reduced case, and James Keenan for reporting it in https://github.com/Perl/perl5/issues/20468.

This should fix the problem.

demerphq committed:
pp_ctl.c - teach module_true how to deal with blocked requires

If require encounters an @INC hook that blocks the require it
ends up with a STUB node as the optree for the LEAVEEVAL. This was
causing the module_true logic to segfault. Guarding against the
OP_STUB node fixes the problem.

This affected the tests for IO::Socket::SSL. Thanks to Graham
Knopp for the reduced case, and James Keenan for reporting it
in https://github.com/Perl/perl5/issues/20468.

This should fix the problem.

demerphq committed:
t/comp/require.t - test module_true with requiring empty files

jkeenan commented:
With this branch I can install both IO-Socket-SSL and Module-Pluggable in the Linux and FreeBSD environments mentioned in https://github.com/Perl/perl5/issues/20468.  (Someone other than me should review the C-code.)  Thanks.

demerphq committed:
t/comp/require.t - add a way to disable the sleep 20

it makes developing the test script quite annoying.

demerphq committed:
t/comp/require.t - add tests for module_true being off

Or the file being empty, or various other cases. We missed
a case in the earlier work on module_true.

This also includes some tweaks to make the tests pass under strict
if needed.

demerphq commented:
fixed up the require tests hopefully
